open_dread_rando/door_patcher.py

```diff
@@ -1,59 +1,309 @@
+_Door049 = {"scenario": "s010_cave", "layer": "default", "actor": "Door049 (PR-PR)"}
+_Door008 = {"scenario": "s010_cave", "layer": "default", "actor": "Door008 (PW-PW)"}
 
-    life_comp = actor.pComponents.LIFE
-    if need_shield:
-        raise NotImplementedError("Adding shields to doors is not implemented")
-    else:
+class DoorPatcher:
+    """An API to patch doors. Call patch_door() to use. """
+    def __init__(self, editor: PatcherEditor) -> None:
+        # get actors from reference dicts
+        self.editor = editor
+        self.SHIELD = editor.resolve_actor_reference(_EXAMPLE_SHIELD)
+        self.PHANTOMDOOR = editor.resolve_actor_reference(_EXAMPLE_PHANTOM_DOOR)
+
+    def door_actor_to_type(self, door: Container, scenario: str) -> DoorType:
+        """
+        find's a door's DoorType
+        creates a list of all doortype enums, and filters them by various means until there is one remaining
+        raises an exception if door is not a patchable type
+        """
+        # filter enums by door's actorref
+        door_actor_ref = door.oActorDefLink.split(':')[1]
+        possible_enum_values = list()
+        for type in DoorType:
+            if door_actor_ref == type.door.actordef:
+                possible_enum_values.append(type)
+        
+        if len(possible_enum_values) == 1:
+            return possible_enum_values[0]
+        
+        # check for shields
+        shield_actor = None
+        life_comp = door.pComponents.LIFE
+        for link_name in ["wpLeftDoorShieldEntity", "wpRightDoorShieldEntity"]:
+            if life_comp[link_name] != '{EMPTY}':
+                shield_actor = self.editor.resolve_actor_reference(self.editor.reference_for_link(life_comp[link_name], scenario))
+        
+        if shield_actor is not None:
+            # remove non shielded doors if shield exists
+            possible_enum_values = list(filter(lambda e: (e.need_shield is True), possible_enum_values))
+
+            # check shields
+            shield_actor_ref = shield_actor.oActorDefLink.split(':')[1]
+            possible_enum_values = list(filter(lambda e: (e.shield.actordef == shield_actor_ref), possible_enum_values))
+        else:
+            possible_enum_values = list(filter(lambda e: (e.need_shield is False), possible_enum_values))
+
+        if len(possible_enum_values) == 1:
+            return possible_enum_values[0]
+        else:
+            return DoorType.INVALID
+
+
+    def patch_door(self, door: dict):
+        """
+        Patches a door given a reference. 
+        
+        @param door: A dictionary representing a requested door patch.
+        """
+
+        scenario = door["actor"]["scenario"]
+        door_type = DoorType.get_type(door["door_type"])
+        if door_type is DoorType.INVALID:
+            raise ValueError(f"{door['door_type']} is not a valid type!")
+        map = self.editor.get_scenario_map(scenario)
```

Comment:
rather than passing this to the functions, it's okay to just call this function while inside the other functions later, since they have the scenario string. the editor caches modified files so it won't slow things down to call it multiple times, and it'll make the signatures of the door patcher cleaner